pandas/core/indexes/datetimelike.py

```diff
@@ -121,8 +121,149 @@ def ceil(self, freq):
         return self._round(freq, np.ceil)
 
 
-class DatetimeIndexOpsMixin(object):
+class DatetimeLikeArray(object):
+    """
+    Shared Base/Mixin class for DatetimeArray, TimedeltaArray, PeriodArray
+
+    Assumes that __new__/__init__ defines:
+        _data
+        freq
+    """
+
+    @property
+    def _box_func(self):
+        """
+        box function to get object from internal representation
+        """
+        raise com.AbstractMethodError(self)
+
+    def __iter__(self):
+        return (self._box_func(v) for v in self.asi8)
+
+    @property
+    def values(self):
+        """ return the underlying data as an ndarray """
+        return self._data.view(np.ndarray)
+
+    @property
+    def asi8(self):
+        # do not cache or you'll create a memory leak
+        return self.values.view('i8')
+
+    # ------------------------------------------------------------------
+    # Null Handling
+
+    @property  # NB: override with cache_readonly in immutable subclasses
```

Comment:
Did you have a PR started that made something like a `@maybe_cache_readonly`? That'll look more appealing with this in place, else we'll just be overriding these just to mark them as cached.

Your PR probably did this, but ideally would would have a class attribute that indicates whether the class is immutable, and a single decorator for both.

Reply:
Yah the PR you're thinking of did exactly that.  At the time there was only one property/cache_readonly affected, but this was the motivation.  It'll be easy to revive if it becomes necessary.